sdk/core/azure-core/src/http/url.cpp

```diff
@@ -1,69 +1,271 @@
+    m_fragment = std::string(pos + 1, url.end());
+  }
+}
+
+std::string Url::Decode(const std::string& value)
+{
+  const static std::vector<int> hexTable = []() {
+    std::vector<int> t(256, -1);
+    for (int i = 0; i < 10; ++i)
+    {
+      t[static_cast<std::size_t>('0') + i] = i;
+    }
+    for (int i = 10; i < 16; ++i)
+    {
+      t[static_cast<std::size_t>('A') + i - 10] = i;
+      t[static_cast<std::size_t>('a') + i - 10] = i;
+    }
+    return t;
+  }();
 
-  // Path
-  this->m_path = std::string(start, endOfUrl);
-  auto pathSize = this->m_path.size();
-  if (pathSize > 0)
+  std::string decodedValue;
+  for (std::size_t i = 0; i < value.size();)
   {
-    auto pathLast = pathSize - 1;
-    // remove any slashes from the end
-    for (unsigned long index = 0; index <= pathLast; index++)
+    char c = value[i];
+    if (c == '+')
+    {
+      decodedValue += ' ';
+      ++i;
+    }
+    else if (c == '%')
     {
-      if (this->m_path[pathLast - index] != '/')
+      if (i + 2 >= value.size() || hexTable[value[i + 1]] < 0 || hexTable[value[i + 2]] < 0)
       {
-        this->m_path = this->m_path.substr(0, pathSize - index);
-        break;
+        throw std::runtime_error("failed when decoding url component");
       }
+      int v = (hexTable[value[i + 1]] << 4) + hexTable[value[i + 2]];
+      decodedValue += static_cast<std::string::value_type>(v);
+      i += 3;
+    }
+    else
+    {
+      decodedValue += value[i];
+      ++i;
+    }
+  }
+  return decodedValue;
+}
+
+static const char* unreserved
+    = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~";
+static const char* subdelimiters = "!$&'()*+,;=";
+
+std::string Url::EncodeHost(const std::string& host)
+{
+  return EncodeImpl(host, [](int c) { return c > 127; });
+}
+
+std::string Url::EncodePath(const std::string& path)
+{
+  const static std::vector<bool> shouldEncodeTable = []() {
+    const std::string pathCharacters
+        = std::string(unreserved) + std::string(subdelimiters) + "%/:@";
+
+    std::vector<bool> ret(256, true);
+    for (char c : pathCharacters)
+    {
+      ret[c] = false;
+    }
+    // we also encode % and +
+    ret['%'] = true;
+    ret['+'] = true;
+    return ret;
+  }();
+
+  return EncodeImpl(path, [](int c) { return shouldEncodeTable[c]; });
+}
+
+std::string Url::EncodeQuery(const std::string& query)
+{
+  const static std::vector<bool> shouldEncodeTable = []() {
+    std::string queryCharacters = std::string(unreserved) + std::string(subdelimiters) + "%/:@?";
+
+    std::vector<bool> ret(256, true);
+    for (char c : queryCharacters)
+    {
+      ret[c] = false;
+    }
+    // we also encode % and +
+    ret['%'] = true;
+    ret['+'] = true;
+    // Surprisingly, '=' also needs to be encoded because Azure Storage server side is so strict.
+    // We are applying this function to query key and value respectively, so this won't affect
+    // that = used to separate key and query.
+    ret['='] = true;
```

Comment:
No.

Reply:
well, this is super `storage-related`. We would have trouble if some other service behaves different.
Maybe we will need to create a Url interface with basic/general encoding/decoding algorithms for all services.
Then each service can override the encoding strategy on a derived Url class.
I will review this with Jeff.  Anyway, this can be done later. As of now we only have storage service, so we wouldn't break stuff

Reply:
No, we won't have any trouble. For some characters, we have to encode them. While for others we don't have to. But server can still decode if we do encode them.

Reply:
I think this storage service specific encoding setting could be a concern if other services don't need it encoded. It would be useful for this encoding map to be customizable to avoid pigeonholing the behavior.

Reply:
ok, yes. Maybe encoding too much is not that bad.
Encoding less than expected by service would be problematic.